src/content/posts/en/spark-structured-streaming-es.md

```diff
@@ -0,0 +1,180 @@
+
+Some simple examples can be found in the [Spark documentation](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html)
+But in real life, solutions will be more complex than that, so here we share a simplified solution of what we have done:
+
+
+```python
+# imports
+from datetime import datetime
+from pyspark.sql.types import StructType, StringType, StructField, \
+   DoubleType, BooleanType, IntegerType
+from pyspark.sql import functions as F
+
+# var definition
+hours_window = 24
+sliding_window = 5
+checkpoint_dir = 'hdfs:///checkpoint_folder/'
+
+# helper to create types
+def _get_types(types_dict):
+   return StructType([(StructField(f_name, f_type, True)) for f_name, f_type in types_dict.items()])
+
+# structure types for the data streaming
+LOGLINE_SCHEMA = _get_types({
+   'timestamp': StringType(),
+   'cost': DoubleType(),
+   'clicks': DoubleType(),
+})
+
+# read stream definition
+logline_df = self.spark \
+   .readStream \
+   .schema(LOGLINE_SCHEMA) \
+   .parquet('hdfs:///mydata/*')
+
+# convert timestamp field for window operation
+logline_df = logline_df.withColumn('timestamp', F.to_timestamp('timestamp'))
+
+
+# add watermark definition
+logline_df = logline_df.withWatermark('timestamp', '1 minute')
+
+
+# compute the metrics
+metrics_df = logline_df.groupBy(
+   F.window(
+      logline_df.timestamp, f'{hours_window} hours', 
+      f'{sliding_window} minutes'
+   ),
+   logline_df.campaign_eid
+) \
+.agg(
+   F.sum('cost').alias('cost'),
+   F.sum('clicks').alias('clicks')
+)
+
+# start streaming
+query = metrics_df.writeStream \
+   .option("checkpointLocation", checkpoint_dir) \
+   .foreachBatch(collect_metrics_fn) \
+   .outputMode('append') \
+   .trigger(processingTime=f'{sliding_window} minutes') \
+   .start()
+
+# function called for each batch
+def collect_metrics_fn(df, epoch_id):
+   # do any operations needed in the dataframe
+   df = df.withColumn('timestamp', F.lit(datetime.now()))
+   # output dataframe
+   df.write.mode('overwrite').parquet('hdfs:///output')
+```
+
+In the code we can see that we have called `.readStream`
+to define the source where we are going to read our [parquet](https://databricks.com/glossary/what-is-parquet) files, in this case  our path is `hdfs:///mydata/`.
+In this example we are using the *Hadoop Distributed File System.*
+
+>More information about HDFS:
+https://hadoop.apache.org/docs/r1.2.1/hdfs_design.html
+
+And in this section
+```python
+   F.window(
+      logline_df.timestamp, f'{hours_window} hours', 
+      f'{sliding_window} minutes'
+   )
+``` 
+
+We define a window period of 24 hours using the *timestamp* field, and sliding every 5 minutes.
+
+To start the streaming we need to call the `writeStream` function.
+
+The **checkpoint** folder definition it’s very important, because if our job fails or gets restarted will try first to recover from the checkpointed data without having to process all the data from scratch again. 
```

Comment:
```suggestion
The `checkpoint` folder definition is very important, because if our job fails or gets restarted will try first to recover from the checkpointed data without having to process all the data from scratch again. 
```